src/api/ExchangeApi/mock/DepositApiMock.ts

```diff
@@ -0,0 +1,159 @@
+import BN from 'bn.js'
+import assert from 'assert'
+
+import { getEpoch } from 'utils'
+import { ZERO, BATCH_TIME } from 'constants/'
+
+import { DepositApi, BalanceState } from 'types'
+
+interface BalancesByUserAndToken {
+  [userAddress: string]: { [tokenAddress: string]: BalanceState }
+}
+
+export class DepositApiMock implements DepositApi {
+  private _balanceStates: BalancesByUserAndToken
+
+  public constructor(balanceStates: BalancesByUserAndToken = {}) {
+    this._balanceStates = balanceStates
+  }
+
+  public async getBatchTime(): Promise<number> {
+    return BATCH_TIME
+  }
+
+  public async getCurrentBatchNumber(): Promise<number> {
+    return Math.floor(getEpoch() / BATCH_TIME)
+  }
+
+  public async getSecondsRemainingInBatch(): Promise<number> {
+    return BATCH_TIME - (getEpoch() % BATCH_TIME)
+  }
+
+  public async deposit(userAddress: string, tokenAddress: string, amount: BN): Promise<void> {
+    // Create the balance state if it's the first deposit
+    const currentStateIndex = await this.getCurrentBatchNumber()
+    let balanceState = this._initBalanceState(userAddress, tokenAddress, currentStateIndex)
+
+    // Update any unapplied deposit
+    this._updateDepositsBalance(userAddress, tokenAddress, currentStateIndex)
+
+    const pendingDeposits = balanceState.pendingDeposits
+    pendingDeposits.amount = pendingDeposits.amount.add(amount)
+    pendingDeposits.stateIndex = currentStateIndex
+  }
+
+  public async requestWithdraw(userAddress: string, tokenAddress: string, amount: BN): Promise<void> {
+    const currentStateIndex = await this.getCurrentBatchNumber()
+    let balanceState = this._initBalanceState(userAddress, tokenAddress, currentStateIndex)
+
+    balanceState.pendingWithdraws = {
+      amount,
+      stateIndex: currentStateIndex,
+    }
+  }
+
+  public async withdraw(userAddress: string, tokenAddress: string): Promise<void> {
+    const currentStateIndex = await this.getCurrentBatchNumber()
+    let balanceState = this._initBalanceState(userAddress, tokenAddress, currentStateIndex)
+
+    const pendingWithdraws = balanceState.pendingWithdraws
+
+    // Check that the user has balance, and the proper request
+    //  The restrictions are stronger than in the smart contract
+    assert(!pendingWithdraws.amount.isZero(), "There wasn't any previous pending withdraw")
+    assert(!balanceState.balance.isZero(), "There user doesn't have any balance")
+    assert(pendingWithdraws.stateIndex < currentStateIndex, 'The withdraw request is not settled yet')
+
+    const amount = BN.min(pendingWithdraws.amount, balanceState.balance)
+    pendingWithdraws.amount = ZERO
+    balanceState.balance = balanceState.balance.sub(amount)
```

Comment:
nit: Could it be simplified to `balanceState.balance -= amount`?
Unless this is a specific TS construct. In that case just ignore this comment.

Reply:
This is a method of BN.js, and unfortunately there's no operator overloading that I'm aware of

Reply:
Got it.